### ISSUE TITLE
Can you automate filter subtype to an incorrect type?

### ISSUE THREAD

**Issue by baconpaul** [surge-synthesizer/surge#2414]
With the merge of #2411 we realized we don't know what stops daw automation from making a filter subtype invalid. That change doesn't make it any different, but we do need to look for 1.7.2 and either find the code that puts in the constraint, or add the code that puts in the constraint

The code would look a lot like the limit range in SurgePatch, just applied in the DAW setParameter codepath also.

**Comment by amyfurniss** [comment]
I experimented with this in Logic and it looks like the val_max in parameter.cpp is limiting the range of automation to valid values.

I can't automate either the filter subtype or filter type beyond valid values in Logic. If I increase the val_max.i in Parameter.cpp for ct_filtertype or ct_filtersubtype it's then possible to enter invalid values. When I enter an invalid value for the main filter type it appears that there is no filter applies (but the synth doesn't crash). It does crash if I enter an invalid subtype.

**Comment by baconpaul** [comment]
Yeah Val_max.i will be the top. The question here is that some filters (like bandpass) have 6 subtypes nad others (like LP4) have 3 subtypes. So even though 6 is the max subtype for any filter, if you have filter set to one which only has 3 subtypes and you set subtype to 4 via automation does it max the subtype to the type specific range.

As you change filter types you’ll see that the RMB on the subtype widget gets a limited list (that’s new!) but does that limited list also limit automation

I think the trick is to not change val_max.i - the DAW won’t be stable under that parameter metadata change - but rather at runtime somewhere do a check of if val > n_filtersyubtypes then (do something reasonable). Which we might have. But this issue is a reminder to check it

The fact that we don’t crash is very very hopeful that the code is there somewhere 👍

**Comment by mkruselj** [comment]
OK so what happens here is that when you automate the subtype parameter, first there's no UI update happening so you don't see that it actually changed.

Second it does indeed go above the number of defined subtypes for a particular filter type.

**Comment by baconpaul** [comment]
I thought it would be as easy as adding the item to nonmod_params for the automation to show up but the switch multi value type is trickier than that. I couldn't get here but here's a good start

Then there's the question of rejecting the set value. That has to happen in the synth somewhere, probably somewhere clever in SurgeSynthesizer::setParameterAutomated. Still pondering that.

Here's the patch fro 10 minutes of hacking this am that almost got automation working.

[patch-2414.txt](https://github.com/surge-synthesizer/surge/files/5062648/patch-2414.txt)
